File: 10.Binary Search/DAY 2/Kth Smallest Number in Multiplication Table.java
package Week10_Binary_Search;

public class Day2_Multiplication_Table {
	 public static int findKthNumber(int m, int n, int k) {
		 int left = 0;
		 int right = m*n;
	     int answer = 0;
	        
	     while(left < right){
	    	 int mid = ( left + right ) / 2;
//	    	 int cnt = getCount(mid, m,n);
	         int count = 0;
	         for(int i = 1 ; i <= m ; i++){
	        	 count += Math.min(mid/i, n); 
	         }//갯수를 구하는 것이 논제!
	         if(count >= k) right = mid;
	         else {
	             left = mid+1;
	         }
	     }
	     return left;

	}
//	public static int getCount(int mid, int m, int n) {
//		int cnt = 0;
//		for(int i = 1 ; i <= m ; i++) {
//			for(int j = 1; j <= n ; j++) {
//				if(mid >= i*j) cnt++;
//				else break;
//			}
//		}
//		return cnt;
//	}
	 
	public static void main(String[] args) {

		int m = 45;
		int n = 12;
		int k = 471;
		System.out.println(findKthNumber(m,n,k));
		

	}

}

File: 10.Binary Search/DAY 2/Kth Smallest Number in Multiplication Table.py
from bisect import bisect_left

# def findKthNumber(m: int, n: int, k: int) -> int:
#     list = []
#
#     for i in range(m):
#         for j in range(n):
#             index = bisect_left(list, (i + 1) * (j + 1))
#             list.insert(index, (i + 1) * (j + 1))
#
#     print(list)
#
#     return list[k - 1]

# def findKthNumber(m: int, n: int, k: int) -> int:
#     dic = {}
#     for i in range(m):
#         for j in range(n):
#             dic[(i + 1) * (j + 1)] = dic.get((i + 1) * (j + 1), 0) + 1
#
#     list = sorted(dic.items())
#
#     print(list)
#
#     for key, values in list:
#         k -= values
#         if k <= 0:
#             return key
#
#     return 0

def findKthNumber(m: int, n: int, k: int) -> int:

    left = 1
    right = m * n

    while left <= right:
        mid = (left + right) // 2
        cnt = 0
        height = 1

        while height <= mid and height <= m:
            cnt += mid // height if mid // height <= n else n
            height += 1

        if cnt <= k:
            left = mid + 1
        else:
            right = mid - 1


    return right

m = 5
n = 6
k = 20

print(findKthNumber(m, n, k))